src/hockeypuck/pghkp/storage.go

```diff
@@ -463,50 +652,511 @@ func (st *storage) insertKeyTx(tx *sql.Tx, key *openpgp.PrimaryKey) (isDuplicate
+
+	// Send all keys to in-mem tables to the pg server; *no constraints checked*
+	err = st.bulkInsertSendBunchTx(keystmt, "keys", keysValueArgs)
+	if err != nil {
+		return err
+	}
+
+	// Send all subkeys to in-mem tables to the pg server; *no constraints checked*
+	err = st.bulkInsertSendBunchTx(subkeystmt, "subkeys", subkeysValueArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type KeyInsertArgs struct {
+	RFingerprint *string
+	jsonStrDoc   *string
+	//	ctime *Time
+	//	mtime *Time
+	MD5      *string
+	keywords *string
+}
+type SubkeyInsertArgs struct {
+	keyRFingerprint    *string
+	subkeyRFingerprint *string
+}
+
+// Insert keys & subkeys to in-mem tables with no constraints at all: should have no errors!
+func (st *storage) bulkInsertDoCopy(keyInsertArgs []KeyInsertArgs, subkeyInsertArgs [][]SubkeyInsertArgs,
+	result *hkpstorage.InsertError) (ok bool) {
+	lenKIA := len(keyInsertArgs)
+	for totKeyArgs, totSubkeyArgs, idx, lastIdx := 0, 0, 0, 0; idx < lenKIA; totKeyArgs, totSubkeyArgs, lastIdx = 0, 0, idx {
+		keysValueStrings := make([]string, 0, 5000)         // could be 10K keys but 5K more uniform for 15K files
+		keysValueArgs := make([]interface{}, 0, 5000*6)     // *** less than 64k arguments ***
+		subkeysValueStrings := make([]string, 0, 32000)     // max 32K subkeys
+		subkeysValueArgs := make([]interface{}, 0, 32000*2) // *** less than 64k arguments ***
+		insTime := make([]time.Time, 0, 5000)               // stupid but anyway...
+		for i, j := 0, 0; idx < lenKIA; idx, i = idx+1, i+1 {
+			lenSKIA := len(subkeyInsertArgs[idx])
+			totKeyArgs += 6
+			totSubkeyArgs += 2 * lenSKIA
+			if (totKeyArgs > 30000) || (totSubkeyArgs > 64000) {
+				totKeyArgs -= 6
+				totSubkeyArgs -= 2 * lenSKIA
+				break
+			}
+			keysValueStrings = append(keysValueStrings,
+				fmt.Sprintf("($%d::TEXT, $%d::JSONB, $%d::TIMESTAMP, $%d::TIMESTAMP, $%d::TEXT, to_tsvector($%d))",
+					i*6+1, i*6+2, i*6+3, i*6+4, i*6+5, i*6+6))
+			insTime = insTime[:i+1] // re-slice +1
+			insTime[i] = time.Now().UTC()
+			keysValueArgs = append(keysValueArgs, *keyInsertArgs[idx].RFingerprint, *keyInsertArgs[idx].jsonStrDoc,
+				insTime[i], insTime[i], *keyInsertArgs[idx].MD5, *keyInsertArgs[idx].keywords)
+
+			for sidx := 0; sidx < lenSKIA; sidx, j = sidx+1, j+1 {
+				subkeysValueStrings = append(subkeysValueStrings, fmt.Sprintf("($%d::TEXT, $%d::TEXT)", j*2+1, j*2+2))
+				subkeysValueArgs = append(subkeysValueArgs,
+					*subkeyInsertArgs[idx][sidx].keyRFingerprint, *subkeyInsertArgs[idx][sidx].subkeyRFingerprint)
+			}
+		}
+		// TODO: remove this & lastIdx (debuging only)
+		log.Infof("Attempting bulk insertion of %d keys and a total of %d subkeys!", idx-lastIdx, totSubkeyArgs>>1)
+		keystmt := fmt.Sprintf("INSERT INTO %s (rfingerprint, doc, ctime, mtime, md5, keywords) VALUES %s",
+			keys_copyin_temp_table_name, strings.Join(keysValueStrings, ","))
```

Comment:
Consider using CopyIn as an alternative to dynamic SQL, which can be risky. See https://pkg.go.dev/github.com/lib/pq#CopyIn and an example, https://pkg.go.dev/github.com/lib/pq?utm_source=godoc#hdr-Bulk_imports.

Reply:
That was my first thought, following [this answer](https://stackoverflow.com/a/41664644) to the batch-sql-statements question. I think I preferred the "INSERT... VALUES" query because it allowed me to use the string originally used for insertion:
"($%d::TEXT, $%d::JSONB, $%d::TIMESTAMP, $%d::TIMESTAMP, $%d::TEXT, to_tsvector($%d))"

`CopyIn(..)` would not let me insert qualifiers like "::JSONB", I think, making them part of the string input. I also doubt I would be able to use `tsvector(...)`, since `CopyIn(..)` creates a "COPY ... FROM STDIN" query.